src/datachain/query/dataset.py

```diff
@@ -682,7 +675,15 @@ def q(*columns):
             names = {c.name for c in columns}
             # Columns for the generated table.
             cols = [c for c in udf_table_cols if c.name in names]
-            return sqlalchemy.select(*cols).select_from(udf_table_query)
+            res = sqlalchemy.select(*cols).select_from(udf_table_query)
+
+            if query._order_by_clauses:
+                # if we are not selecting all rows in UDF, we need to ensure that
+                # we get the same rows as we got as inputs of UDF since selecting
+                # without ordering can be non deterministic in some databases
+                res = res.order_by(udf_table_query.c.sys__id)
```

Comment:
Move `order_by` here, [same](https://github.com/iterative/datachain/blob/main/src/datachain/query/dataset.py#L633-L636) as in `UDFSignal.create_result_query` method.

Reply:
I am not against this change but PTAL at this: https://github.com/iterative/datachain/issues/477#issuecomment-2387274525. It looks unlikely we'll be able to support this behaviour in BQ.